src/coreclr/System.Private.CoreLib/src/Internal/Runtime/InteropServices/InMemoryAssemblyLoader.cs

```diff
@@ -2,48 +2,87 @@
+        /// into the specified load context.
+        /// </summary>
+        /// <param name="moduleHandle">The native module handle for the assembly.</param>
+        /// <param name="assemblyPath">The path to the assembly (as a pointer to a UTF-16 C string).</param>
+        /// <param name="loadContext">Load context (currently must be IntPtr.Zero)</param>
+        [UnmanagedCallersOnly]
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
+            Justification = "The same C++/CLI feature switch applies to LoadInMemoryAssembly and this function. We rely on the warning from LoadInMemoryAssembly.")]
+        internal static unsafe void LoadInMemoryAssemblyInContext(IntPtr moduleHandle, IntPtr assemblyPath, IntPtr loadContext)
+        {
+            if (!IsSupported)
+                throw new NotSupportedException("This API is not enabled in trimmed scenarios. see https://aka.ms/dotnet-illink/nativehost for more details");
+
+            if (loadContext != IntPtr.Zero)
+                throw new ArgumentOutOfRangeException(nameof(loadContext));
+
+            LoadInMemoryAssemblyInContextImpl(moduleHandle, assemblyPath, AssemblyLoadContext.Default);
+        }
+
+        [RequiresUnreferencedCode("C++/CLI is not trim-compatible", Url = "https://aka.ms/dotnet-illink/nativehost")]
+        private static void LoadInMemoryAssemblyInContextImpl(IntPtr moduleHandle, IntPtr assemblyPath, AssemblyLoadContext? alc = null)
+        {
             string? assemblyPathString = Marshal.PtrToStringUni(assemblyPath);
             if (assemblyPathString == null)
-            {
                 throw new ArgumentOutOfRangeException(nameof(assemblyPath));
-            }
 
-            // We don't cache the ALCs here since each IJW assembly will call this method at most once
+            // We don't cache the ALCs or resolvers here since each IJW assembly will call this method at most once
             // (the load process rewrites the stubs that call here to call the actual methods they're supposed to)
-#pragma warning disable IL2026 // suppressed in ILLink.Suppressions.LibraryBuild.xml
-            AssemblyLoadContext context = new IsolatedComponentLoadContext(assemblyPathString);
-#pragma warning restore IL2026
-            context.LoadFromInMemoryModule(moduleHandle);
-#else
-            throw new PlatformNotSupportedException();
-#endif
+            if (alc is null)
+            {
+                alc = new IsolatedComponentLoadContext(assemblyPathString);
+            }
+            else if (alc == AssemblyLoadContext.Default)
+            {
+                var resolver = new AssemblyDependencyResolver(assemblyPathString);
+                AssemblyLoadContext.Default.Resolving +=
+                    [RequiresUnreferencedCode("C++/CLI is not trim-compatible", Url = "https://aka.ms/dotnet-illink/nativehost")]
```

Comment:
I had marked this `UnconditionalSuppressMessage`, since that seemed more in line with my intent, but was still getting warnings from the analyzer (I think?). Is that expected or something weird with lambdas?

Reply:
There's a bug for this in roslyn: https://github.com/dotnet/roslyn/issues/59746